src/MorphoScripts.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
+import {SafeERC20} from "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import {IMorpho, MarketParams, Position} from "src/interfaces/IMorpho.sol";
+import {IMetaMorpho} from "src/interfaces/IMetaMorpho.sol";
+import {IMorphoUniversalRewardsDistributor} from "src/interfaces/IMorphoUniversalRewardsDistributor.sol";
+import {DeFiScriptErrors} from "src/lib/DeFiScriptErrors.sol";
+
+contract MorphoVaultActions {
+    // To handle non-standard ERC20 tokens (i.e. USDT)
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Deposit assets into a MetaMorpho vault
+     * @param vault The address of the MetaMorpho vault
+     * @param asset The address of the asset to deposit
+     * @param amount The amount of the asset to deposit
+     */
+    function deposit(address vault, address asset, uint256 amount) external {
+        IERC20(asset).forceApprove(vault, amount);
+        IMetaMorpho(vault).deposit({assets: amount, receiver: address(this)});
+    }
+
+    /**
+     * @notice Withdraw assets from a MetaMorpho vault
+     * As suggested from MetaMorpho.sol doc, it is recommended to not use their
+     * redeemMax function to retrieve max shares to redeem due to cost.
+     * Instead will just use balanceOf(vault) to optimistically redeem all shares when amount is `type(uint256).max`.
+     * @param vault The address of the MetaMorpho vault
+     * @param amount The amount of assets to withdraw, if it is `type(uint256).max`, it will withdraw max
+     */
+    function withdraw(address vault, uint256 amount) external {
+        if (amount == type(uint256).max) {
+            // Withdraw max
+            IMetaMorpho(vault).redeem({
+                shares: IMetaMorpho(vault).balanceOf(address(this)),
+                receiver: address(this),
+                owner: address(this)
+            });
+        } else {
+            IMetaMorpho(vault).withdraw({assets: amount, receiver: address(this), owner: address(this)});
+        }
+    }
+}
+
+contract MorphoActions {
+    // To handle non-standard ERC20 tokens (i.e. USDT)
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Repay assets and withdraw collateral from a Morpho blue market on behalf of `onBehalf` and send collateral to `receiver`
+     * @param morpho The address of the top level Morpho contract
+     * @param marketParams The market parameters of the individual morpho blue market
+     * @param repayAmount The amount of assets to repay, pass in `type(uint256).max` to repay max
+     * @param repayShares The amount of shares to repay
+     * @param withdrawAmount The amount of assets to withdraw as collateral
+     */
+    function repayAndWithdrawCollateral(
+        address morpho,
+        MarketParams memory marketParams,
+        uint256 repayAmount,
+        uint256 repayShares,
```

Comment:
we abstract over shares in every other action method, but not this one -- is there a particular reason why we expose the shares concept for `repayAndWithdrawCollateral`?

It seems from  a straightforward read of the code that there's one path that cares about `repayShares`, where we do not specify an exact asset amount to repay. But from my understanding of morpho, the only likely case where we would specify shares is repay max -- and we encode that using `type(uint256).max` amounts.

As it stands, we are also not doing anything to catch it if a caller specifies both amounts. This is ambiguous, and probably an error? Ideally it shouldn't even be possible to specify the same amount two ways, since it creates these invalid states.

Looking at the builder code in #67, I see that we always encode the calldata with `0` as repay shares. (In fact 0 is hardcoded.) So it would appear that this codepath is unused in practice. I see we have one test-case for repay-with-shares, but this is actually just the "repay max" test scenario, which is already covered by the `type(uint256).max` case.

Should we not just abstract over shares entirely, and remove `repayShares` from this method?

Reply:
And yes we don't use repayShares now, as we only want to use it when we want to repay max. 
Since scripts handles repay max we don't need to use this field anymore. 

But leave repayShares here as MorphoScripts is just scripts to interact with Morpho, so since they open that options up, so I didn't completely trim off the whole interfaces they had. 

But agree, we can just remove it.


For both amounts, I just passed it to Morpho contract and let them handle it. 
As far as how Morpho handles it  is (amounts over shares / shares over amount) and won't throw error.

Reply:
> But agree, we can just remove it.

I think that's best, unless we're actually going to use shares-based accounting in some way.

> For both amounts, I just passed it to Morpho contract and let them handle it. As far as how Morpho handles it is (amounts over shares / shares over amount) and won't throw error.

Yeah, but we have an opportunity abstracting over Morpho to avoid situations like this, where there's an ambiguous or invalid state. :) Not an issue, of course, if we just support `amount` and forget about shares.

Reply:
Sounds good  👍